File: frontend/src/app/store/chat-store/service/chat.service.ts
import {Injectable} from '@angular/core';
import {environment} from "../../../../environments/environment";
import {HttpClient} from "@angular/common/http";
import {ChatRequest} from "./chat.request";
import {Chat} from "./chat.model";

@Injectable({
  providedIn: 'root'
})
export class ChatService {
  private apiUrl = environment.baseUrl + "/api/chat"

  constructor(private http: HttpClient) {
  }

  createChat(request: ChatRequest) {
    let formData = new FormData()
    if (request.picture)
      formData.append('picture', request.picture)
    formData.append('name', request.name)
    formData.append('description', request.description)
    return this.http.post<Chat>(this.apiUrl, formData)
  }

  updateChat(chatId: number | string, request: ChatRequest) {
    let formData = new FormData()
    if (request.picture)
      formData.append('picture', request.picture)
    formData.append('name', request.name)
    formData.append('description', request.description)
    return this.http.put<Chat>(this.apiUrl + '/' + chatId, formData)
  }

  getChatsNearby() {
    return this.http.get<Chat[]>(this.apiUrl + '/nearby')
  }

  getUserChats() {
    return this.http.get<Chat[]>(this.apiUrl + '/user')
  }

  getChat(chatId: number | string | null) {
    return this.http.get<Chat>(this.apiUrl + '/' + chatId)
  }

  joinChat(chatId: number | string) {
    return this.http.post<void>(this.apiUrl + '/join', {chatId: chatId})
  }

  leaveChat(id: number | string) {
    return this.http.post<void>(this.apiUrl + '/leave', {chatId: id})
  }

  deleteChat(id: number | string) {
    return this.http.delete<void>(this.apiUrl + '/' + id)
  }

  enableNotification(chatId: number | string) {
    return this.http.post<void>(this.apiUrl + '/notification/enable', {chatId: chatId})
  }

  disableNotification(chatId: number | string) {
    return this.http.post<void>(this.apiUrl + '/notification/disable', {chatId: chatId})
  }
}

File: backend/src/main/java/org/example/backend/service/ChatService.java
package org.example.backend.service;

import org.example.backend.model.entity.Chat;
import org.example.backend.model.entity.Message;
import org.example.backend.model.entity.User;
import org.example.backend.model.request.SaveChatRequest;

import java.util.List;

public interface ChatService {
    List<Chat> getUserChats();

    List<Chat> getChatsNearby();

    Chat saveChat(SaveChatRequest request);

    Chat saveChat(Chat chat);

    Chat getChat(Long id);

    Message joinChat(Long chatId);

    Message leaveChat(Long chatId);

    void deleteChat(Long chatId);

    Chat updateChat(Long chatId, SaveChatRequest request);

    void addUser(Chat chat, User user);

    void removeUser(Chat chat, User user);
}
